File: ISVT.m
clc; clear all; close all;
% Creating random matrix data
n=500;
d=10;
P=randn(n,d)
Q=randn(n,d);
X=P*Q'; % 500 x 500 matrix

% samples
sampfrac=.5;
m=floor(sampfrac*n^2);
%observed 
omega = randsample(n^2, m);% returns m x1 vector - uniform dist over 1 to n^2

% observations
Y = zeros(n);

Y(omega)=X(omega);

Y0 = Y; %Saving incomplete matrix

% Using ISVT - making rank 10
T = ceil(50/sampfrac) % Running the loop T times
error = zeros(1,T);
for i=1:T
    [u, s, v] = svd(Y);
    u=u(:,1:d);    
    s=s(1:d,1:d);
    v=v(:,1:d);
    Y=u*s*v';
    Y(omega)=X(omega);
    
    error(i) = norm(Y-X,'fro');
    if(error(i) < 0.01)
        error(i)
        fprintf('Converged at %d \n',i);
        break;
    end
end

File: isvt.py
import numpy as np
import random

m, n = 100, 10000
omega_size = 1000000
num_iter = 20

X = np.random.randn(m, n)
print np.linalg.matrix_rank(X)

omega = zip([random.randint(0,m - 1) for _ in xrange(omega_size)], 
			[random.randint(0,n - 1) for _ in xrange(omega_size)])

X_hat = np.zeros(shape=X.shape)
for x, y in omega:
	X_hat[x, y] = X[x, y]

r = 80
for _ in range(num_iter):
	U, S, V = np.linalg.svd(X_hat, full_matrices=False)
	X_hat = np.dot(np.dot(U[:, :r], np.diag(S[:r])), V[:r, :])
	for x, y in omega:
		X_hat[x, y] = X[x, y]
	print np.linalg.norm(X - X_hat, ord='fro')/np.linalg.norm(X, ord='fro')
